src/api/environment.cc

```diff
@@ -87,10 +87,36 @@ static void HostCleanupFinalizationGroupCallback(
 }
 
 void* NodeArrayBufferAllocator::Allocate(size_t size) {
+  void* ret;
   if (zero_fill_field_ || per_process::cli_options->zero_fill_all_buffers)
-    return UncheckedCalloc(size);
+    ret = UncheckedCalloc(size);
   else
-    return UncheckedMalloc(size);
+    ret = UncheckedMalloc(size);
+  if (LIKELY(ret != nullptr))
+    total_mem_usage_ += size;
+  return ret;
+}
+
+void* NodeArrayBufferAllocator::AllocateUninitialized(size_t size) {
+  void* ret = node::UncheckedMalloc(size);
+  if (LIKELY(ret != nullptr))
+    total_mem_usage_ += size;
+  return ret;
+}
+
+void* NodeArrayBufferAllocator::Reallocate(
+    void* data, size_t old_size, size_t size) {
+  void* ret = UncheckedRealloc<char>(static_cast<char*>(data), size);
+  if (LIKELY(ret != nullptr) || UNLIKELY(size == 0)) {
+    total_mem_usage_ += size;
+    total_mem_usage_ -= old_size;
```

Comment:
Sorry to be a nitpicker but this will emit multiple (heavy-weight) memory barriers on most architectures.

Now that you've changed the type to size_t, it's safe again to write it as `total_mem_usage_ += size - old_size;`. You could add a static_assert to verify that the type sizes match up.

A further optimization is to call fetch_add() with `std::memory_order::relaxed` since it doesn't need to synchronize with anything. The default ordering, seq_cst, emits a full memory barrier, even on x86_64.

Reply:
> Sorry to be a nitpicker but this will emit multiple (heavy-weight) memory barriers on most architectures.

No need to apologize, you’re right :)

I kind of assumed that the memory barriers wouldn’t show up in the overall cost of allocating `ArrayBuffer`s, but I’ve also switched to `std::memory_order::relaxed` for accesses here.